### PR TITLE
CI: Check types more thoroughly with mypy (#295)

### PR DESCRIPTION
**Pull request by SiQube**
ass discussed in PR #293 we add 

```toml
[tools.mypy]
check_untyped_defs = true
disallow_any_generics = true
disallow_incomplete_defs = true
disallow_untyped_defs = true
warn_redundant_casts = true
warn_unused_ignores = true
```

to pyproject.toml. 

all code was adjusted accordingly. in some places I didn't figure out what exactly the type is and had to insert Any -- might change this later

### CODE REVIEW

File: docs/__init__.py
Comment (dkrako) [inline docs/__init__.py:1]: is this file required?

File: src/pymovements/datasets/dataset.py [hunk 1]
```diff
@@ -432,7 +432,9 @@ def pix2deg(self, verbose: bool = True) -> Dataset:
 
         return self
 
-    def pos2vel(self, method: str = 'smooth', verbose: bool = True, **kwargs) -> Dataset:
+    def pos2vel(
+            self, method: str = 'smooth', verbose: bool = True, **kwargs: int | float | str,
```
Comment (dkrako) [inline src/pymovements/datasets/dataset.py:436]: If the function signature takes up more than a line I personally prefer to have each argument on a separate line.

File: src/pymovements/datasets/dataset.py [hunk 2]
```diff
@@ -473,7 +475,7 @@ def detect_events(
             eye: str | None = 'auto',
             clear: bool = False,
             verbose: bool = True,
-            **kwargs,
+            **kwargs: int,
```
Comment (dkrako) [inline src/pymovements/datasets/dataset.py:478]: can you use `float | int | str` here? then it's equal to `pymovements.events.EventDetectionCallable`

File: src/pymovements/events/events.py [hunk 3]
```diff
@@ -125,15 +125,15 @@ def schema(self) -> pl.type_aliases.SchemaDict:
     def __len__(self) -> int:
         return self.frame.__len__()
 
-    def __getitem__(self, *args, **kwargs) -> Any:
+    def __getitem__(self, *args: str, **kwargs: str) -> Any:
```
Comment (dkrako) [inline src/pymovements/events/events.py:128]: the `__getitem__()` method from `polars.DataFrame` is overloaded with a range of different types:
https://github.com/pola-rs/polars/blob/master/py-polars/polars/dataframe/frame.py#L1467-L1657

`polars` is actually a bit confusing when used with this traditional bracket indexing.
`df['a']` will get you the column 'a' as a series, `df[1]` will give you the first row as a dataframe and
`df[1, 'a']` will give you the value of the specific cell.

maybe we should use the following signature to be a bit more  consistent with the polars interface?
```
    def __getitem__(
        self,
        item: str | int | np.ndarray[Any, Any] | tuple[int, int] | tuple[int, str]),
    ) -> Any:
```

 or the full deal by supporting slices as indices:

```
MultiRowSelector: TypeAlias = slice | range | list[int] | pl.Series
MultiColSelector: TypeAlias = slice | range | list[int] | list[str] | list[bool] | pl.Series

def __getitem__(
        self,
        item: (
            str
            | int
            | np.ndarray[Any, Any]
            | MultiColSelector
            | tuple[int, MultiColSelector]
            | tuple[MultiRowSelector, MultiColSelector]
            | tuple[MultiRowSelector, int]
            | tuple[MultiRowSelector, str]
            | tuple[int, int]
            | tuple[int, str]
        ),
    ) -> DataFrame | Series:
```

File: src/pymovements/utils/checks.py [hunk 4]
```diff
@@ -94,7 +95,9 @@ def check_shapes_positions_velocities(positions: np.ndarray, velocities: np.ndar
         )
 
 
-def check_two_kwargs(**kwargs) -> None:
+def check_two_kwargs(
```
Comment (dkrako) [inline src/pymovements/utils/checks.py:98]: I think this can be a single line

File: src/pymovements/utils/checks.py [hunk 5]
```diff
@@ -111,7 +114,9 @@ def check_two_kwargs(**kwargs) -> None:
         raise ValueError('there must be exactly two keyword arguments in kwargs')
 
 
-def check_is_mutual_exclusive(**kwargs) -> None:
+def check_is_mutual_exclusive(
```
Comment (dkrako) [inline src/pymovements/utils/checks.py:117]: I would use `Any` here, as the type of the kwarg really does not matter

File: src/pymovements/utils/checks.py [hunk 6]
```diff
@@ -137,7 +142,9 @@ def check_is_mutual_exclusive(**kwargs) -> None:
         )
 
 
-def check_is_none_is_mutual(**kwargs) -> None:
+def check_is_none_is_mutual(
+        **kwargs: None | list[int] | np.ndarray,
```
Comment (dkrako) [inline src/pymovements/utils/checks.py:146]: I would use `Any` here, as the type of the kwarg really does not matter

File: src/pymovements/utils/checks.py [hunk 7]
```diff
@@ -164,7 +171,10 @@ def check_is_none_is_mutual(**kwargs) -> None:
         )
 
 
-def check_is_length_matching(**kwargs) -> None:
+def check_is_length_matching(
+        **kwargs: np.ndarray | list[list[float]] | list[tuple[float, float]] |
```
Comment (dkrako) [inline src/pymovements/utils/checks.py:175]: simply use `Sized` here

```
from typing import Sized
```

File: src/pymovements/utils/checks.py [hunk 8]
```diff
@@ -183,5 +193,10 @@ def check_is_length_matching(**kwargs) -> None:
     value_1 = kwargs[key_1]
     value_2 = kwargs[key_2]
 
+    # mypy does not understand value_1 and value_2 can't be None
```
Comment (dkrako) [inline src/pymovements/utils/checks.py:196]: will this still be needed if `None` is not one of the accepted types (is not `Sized`)?

File: src/pymovements/utils/decorators.py [hunk 9]
```diff
@@ -25,16 +25,16 @@
 from typing import Any
 
 
-def auto_str(cls):
+def auto_str(cls: Any) -> str:
```
Comment (dkrako) [inline src/pymovements/utils/decorators.py:28]: this is actually a decorator and should return the same class as the input class.

you can achieve this by:
```
ClassT = TypeVar("ClassT") 

def auto_str(cls: ClassT) -> ClassT:
```
see here:
https://stackoverflow.com/a/65613529

In the example they also use the following argument:
`bound=Callable[..., Any]`. I don't know if this is applicable in our case, as this decorator should only be used for classes not functions.

File: src/pymovements/utils/decorators.py [hunk 10]
```diff
@@ -25,16 +25,16 @@
...
     Automatically generate __str__() to include all arguments. Can be used as a decorator.
     """
-    def shorten(value: Any):
+    def shorten(value: Any) -> str:
         if isinstance(value, float):
             value = f'{value:.2f}'
         return value
 
-    def __str__(self):
+    def __str__(self: Any) -> str:
```
Comment (dkrako) [inline src/pymovements/utils/decorators.py:37]: it's a bit strange seing `self ` being typed but is necessary as `auto_str` is a function not a class, right?

**Review by dkrako — CHANGES_REQUESTED** [review]
Thanks a lot for this work!
Seeing the commit history this seems to have been a tough fight with mypy and the external dependencies.

Some small details should be changed though.

Again, thanks a lot for the time you sunk into this.

File: docs/__init__.py
Comment (SiQube) [reply docs/__init__.py:1]: unfortunately otherwise mypy doesn't pick up on it in the pyproject toml file (to ignore typed definitions), another solution would be to type the definitions in `docs/`

File: src/pymovements/datasets/dataset.py (hunk 2 above)
Comment (SiQube) [reply src/pymovements/datasets/dataset.py:478]: I then get:

```console
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/pymovements/datasets/dataset.py:554: error: Argument 4 to "__call__" of "EventDetectionCallable" has incompatible type "**Dict[str, Union[float, int, str]]"; expected "int"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

File: docs/__init__.py
Comment (dkrako) [reply docs/__init__.py:1]: alright, then lets keep it that way

File: src/pymovements/datasets/dataset.py (hunk 2 above)
Comment (dkrako) [reply src/pymovements/datasets/dataset.py:478]: that's very strange, but then lets keep it that way for now and fix it later when trouble comes up.

i think it's because minimum duration is not explicit in this method signature although it's part of the `EventDetectionCallable` signature

File: src/pymovements/datasets/dataset.py (hunk 1 above)
Comment (SiQube) [reply src/pymovements/datasets/dataset.py:436]: done

File: src/pymovements/events/events.py (hunk 3 above)
Comment (SiQube) [reply src/pymovements/events/events.py:128]: done

File: src/pymovements/utils/checks.py (hunk 4 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:98]: done

File: src/pymovements/utils/checks.py (hunk 5 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:117]: done

File: src/pymovements/utils/checks.py (hunk 6 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:146]: done

File: src/pymovements/utils/checks.py (hunk 7 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:175]: we then get

```console
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/pymovements/events/events.py:91: error: Argument "offsets" to "check_is_length_matching" has incompatible type "Union[List[int], ndarray[Any, Any], None]"; expected "Sized"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

File: src/pymovements/utils/checks.py (hunk 8 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:196]: no, but see comment above (https://github.com/aeye-lab/pymovements/pull/295#discussion_r1165202105)

File: src/pymovements/utils/decorators.py (hunk 10 above)
Comment (SiQube) [reply src/pymovements/utils/decorators.py:37]: yea :(

File: src/pymovements/utils/decorators.py (hunk 9 above)
Comment (SiQube) [reply src/pymovements/utils/decorators.py:28]: done

File: src/pymovements/utils/checks.py (hunk 7 above)
Comment (SiQube) [reply src/pymovements/utils/checks.py:175]: I could of course use Any

**Comment by SiQube** [comment]
@dkrako ready for review

File: src/pymovements/events/events.py [hunk 11]
```diff
@@ -88,6 +88,7 @@ def __init__(
 
             # Make sure lengths of onsets and offsets are equal.
             if onsets is not None:
+                assert offsets is not None
```
Comment (dkrako) [inline src/pymovements/events/events.py:91]: Please use `if offsets is None: raise ValueError()` here.

Asserts in production code are problematic and should be used exclusively for stuff that should never happen independent of user input.
If something is failing due to user input error or other dependent systems please use an exception instead.

See this article here: https://www.inspiredpython.com/article/assertions-and-exceptions-are-not-the-same

**Review by dkrako — CHANGES_REQUESTED** [review]
Thanks a lot for this work!

One small change I would like to request: please replace this one assert with raising an exception. Then we're good to go!

File: src/pymovements/events/events.py (hunk 11 above)
Comment (SiQube) [reply src/pymovements/events/events.py:91]: this happens in [line 86-87](https://github.com/aeye-lab/pymovements/pull/295/files#diff-82cc6c96b216cd7ff8cefcac7928eb28601370e2180b3729e129a73e6c31fe7fR86-R87)

```python
# Make sure that if either onsets or offsets is None, the other one is None too.
checks.check_is_none_is_mutual(onsets=onsets, offsets=offsets)
```

Comment (SiQube) [reply src/pymovements/events/events.py:91]: I'll pushed another commit with a comment explicitly pointing at line 87 that it is within mypy and in line 91 there can not be any none offsets

Comment (dkrako) [reply src/pymovements/events/events.py:93]: ah sorry I didn't see that correctly, so everything's fine and I'll merge after the tests complete

Thanks a lot again!

**Review by dkrako — APPROVED** [review]
